Some static compilation tweaks

timholy opened the pull request:
Note this is against #8745. @vtjnash, I made small commits so you can cherry-pick as you wish. I haven't yet changed the `.libs` path, but intend to do so.

The Images tests are working for me. I've been trying this against my own internal code tree, and it works astonishingly well---I've only had to made one restructuring, splitting a file that contained two toplevel modules into two files.

Not all is perfect, of course. Currently, the roadblock is a segfault that can be reproduced with the following sequence:

```
Pkg.add("Interpolations")
Pkg.checkout("Interpolations", "teh/moretypes")  # not sure this is necessary, but it's what I'm using
Base.require(:Interpolations, true)
```

restart julia

```
using Interpolations
A = rand(10);
itp = interpolate(A, BSpline(Constant), OnCell);
itp[1]  # Segfault
```

I will dig into this myself, but I have to take a break for a bit.

timholy committed:
Add missing base_include method

This fixes errors that crop up with multiple workers, e.g.,
ERROR: MethodError: `base_include` has no method matching base_include(::ASCIIString, ::ASCIIString, ::Tuple{Int64,ASCIIString})
Closest candidates are:
  base_include(::Nullable{Union{UTF8String,ASCIIString}}, ::AbstractString, ::Any)
  base_include(::AbstractString, ::Any)
  base_include(::Nullable{Union{UTF8String,ASCIIString}}, ::AbstractString)
  ...
 in eval at sysimg.jl:14
 in anonymous at multi.jl:1303
 in run_work_thunk at multi.jl:584
...

Perhaps you'd prefer a call-site fix?

timholy committed:
Fix serialization warning from workers

timholy commented:
The segfault is happening during a stagedfunction compilation, with a line like this:
```
#1  0x00007ffff6db5802 in array_to_list (a=0x7ffdf6ad8fd0, pv=0x7fffffffae68)
    at ast.c:413
413	        temp = julia_to_scm_(jl_cellref(a,i));
(gdb) call jl_(a)
Array{Any, 1}[:padextract, #<549>, :d]
```
In my time monkeying around with code in `src`, I've not actually seen anything like `#<549>`, so I'm not even sure what this means. I tried this:
```
(gdb) macro define jl_typeof jl_typeof
(gdb) call jl_(jl_typeof(jl_cellref(a,1)))

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff6e6a344 in jl_typeof (v=0x225) at jlapi.c:290
290	    return jl_typeof__MACRO(v);
The program being debugged was signaled while in a function called from GDB.
GDB remains in the frame where the signal was received.
To change this behavior use "set unwindonsignal on".
Evaluation of the expression containing the function
(jl_typeof) will be abandoned.
When the function is done executing, GDB will silently stop.
```
suggesting corrupted memory.

vtjnash commented:
It means something went horribly wrong. It expected a pointer, but got a (small) number (549) instead. 

I suspect Interpolations is the first test case to use staged functions (this PR predates the PR that added staged functions, IIRC)

vtjnash commented on base/deprecated.jl:
```diff
@@ -634,3 +634,15 @@ end
+
+function require(mod::AbstractString)
+    depwarn("`require` is deprecated, use `using` or `import` instead", :require)
+    Base.require(symbol(mod))
```
should this attempt to take the `basename(mod)` and remove the potential `.jl` suffix? or should it be a copy of the old `Base.require(::AbstractString)` logic?

also, `Base.` is unnecessary.

timholy replied:
I think I fixed this. Also added versioning to the libs path.

timholy commented:
I've come up with a simple test case for the segfault. Create the file `Generated.jl` with the following contents:
```julia
module Generated

immutable MyArray{T,N} <: AbstractArray{T,N}
    coefs::Array{T,N}
end

@generated function Base.size{T,N}(A::MyArray{T,N}, d)
    quote
        d <= $N ? size(A.coefs, d) : 1
        # size(A.coefs, d)
    end
end

end
```
and then
```julia
julia> Base.require(:Generated, true)

julia> using Generated

julia> A = Generated.MyArray(rand(10));

julia> size(A, 1)

signal (11): Segmentation fault
```
If you instead use the commented-out definition in `size`, it works fine.

timholy commented:
Even more informative:
```julia
module Generated

immutable MyArray{T,N} <: AbstractArray{T,N}
    coefs::Array{T,N}
end

@generated function five_dims{T,N}(A::MyArray{T,N})
    n = 5
    quote
        $n
    end
end

@generated function n_dims{T,N}(A::MyArray{T,N})
    quote
        $N
    end
end

@generated function Base.length(A::MyArray)
    len = 22
    quote
        $len
    end
end


@generated function Base.ndims{T,N}(A::MyArray{T,N})
    quote
        $N
    end
end

end
```
Only the last one causes a segfault. So it's a pretty narrowly-focused problem.

vtjnash commented:
nice test-case. the issue is that `jl_add_method` now edits the AST of `@generated` functions to add the list sparams. but during precompilation, i call this method twice (once during compilation, and later during reconstruction of the AST), resulting in the list getting duplicated.

fun fact: Jeff made that change 3 days after I made the original PR for this branch (https://github.com/JuliaLang/julia/commit/81136afd7e9899283eb94824995708fdd7245318#diff-53dc66d2cb199a2ea107533c1f99f57bR1778)

JeffBezanson commented:
Hm, maybe that hack is rendered unnecessary by https://github.com/JuliaLang/julia/commit/7908246898641c478901542019e09671ce354bc2#diff-53dc66d2cb199a2ea107533c1f99f57bR924

JeffBezanson commented:
Ah, my mistake, nope. Do you want me to fix this?

timholy committed:
Add `require` deprecations

timholy committed:
Use julia-versioning in libs path

This way people who have two julia installations, release and
master, won't have to recompile their packages all the time.
(Relevant only once 0.4 or higher _is_ release, and master is
something even newer.)

vtjnash commented:
@JeffBezanson sure. i wasn't certain of whether there could be other potential side-effects of that modification. i would probably just write the simple workaround to make this work with minimal changes.
